### PR TITLE
[AMDGPU] Add option to pre-allocate SGPR spill VGPRs

### PR DESCRIPTION
**Pull request by perlfu**
SGPR spill VGPRs are WWM registers so allow them to be allocated by SIPreAllocateWWMRegs pass.
This intentionally prevents spilling of these VGPRs when enabled.

### CODE REVIEW

**Comment by perlfu** [comment]
This is based on #70618 - but I have not included those changes in this PR for clarify.
Note: this is not the only way to a achieve this, we could modify the SGPR spill pass to force phys allocation with a similar amount of code diff.

File: llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.cpp
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.cpp:210]: Don't see why we would want this controllable by an attribute for migration purposes a flag is sufficient.

@cdevadas is also working on getting the proper allocator to handle the allocation of WWM registers

Comment (cdevadas) [reply llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.cpp:210]: For shaders, I hope the SGPR spills are very limited and they won't require more VGPRs causing the 'ran out of allocatable registers' error.

Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.cpp:210]: > Don't see why we would want this controllable by an attribute for migration purposes a flag is sufficient.

For graphics we use the LLVM backed as a shared library so passing options into it  via a command line argument (even if it's just for migration) is a terrible interface.

Comment (perlfu) [reply llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.cpp:210]: As @jayfoad pointed out we typically use attributes as we can start embedding this in IR regardless of backend version.

I have spoken with @cdevadas about more complete allocator handling of WWM registers, and will support migration to it when from the graphics side when it is ready.

In the immediate term we need to work around corruption that split-spilling of SGPR spill VGPRs is causing (in complex divergent control flow) and can take the hit of pre-allocating the VGPRs required for all our current workloads.

**Comment by perlfu** [comment]
Rebased after landing WWM pre-allocation pass move.

**Comment by ruiling** [comment]
I think it is reasonable to let `SIPreAllocateWWMRegs` allocate wwm register given that wwm register allocation by RA is pretty broken. Maybe it is better to make this default behavior. The trick we have done in ` SILowerSGPRSpills::extendWWMVirtRegLiveness()` no longer helps correctness when wwm register is either spilled or its live-range was split. So, can we teach RA to neither spill virtual WWM registers nor split its live-range?

**Comment by cdevadas** [comment]
> `SILowerSGPRSpills::extendWWMVirtRegLiveness()` no longer helps correctness when wwm register is either spilled or its live-range was split. So, can we teach RA to neither spill virtual WWM registers nor split its live-range?

That might not help. I observed for some HPCs and instrumented (asan enabled) code, the number of SGPR spills is quite large eventually requiring more virtual VGPRs while lowering them in SILowerSGPRSpills. These virtual VGPRs are of large live-ranges and often get allocated only after live-range split as their original live-ranges somehow ended up with RegMask interferences with PhysRegs (with the default 64 MaxNumVGPRs available during regalloc). Allocation happens only after splitting them into smaller live-ranges. So, I believe we can't avoid liverange split and spill for wwm-allocation.

**Review by ruiling — COMMENTED** [review]
You may need to check within processDef() whether a virtual register was already pushed into `RegsToRewrite` so we don't need to process one register several times as sgpr spill may have lots of re-definition. Other parts LGTM.

**Comment by ruiling** [comment]
> You may need to check within processDef() whether a virtual register was already pushed into `RegsToRewrite` so we don't need to process one register several times as sgpr spill may have lots of re-definition. Other parts LGTM.

A second look, `VRM->hasPhys(Reg)` should already do that?

**Review by ruiling — APPROVED** [review]
should be good to go if no further concern from others.

**Review by cdevadas — APPROVED** [review]
LGTM

**Pull request merged**
